include/circt/Dialect/RTL/RTLCombinatorial.td

```diff
@@ -200,14 +200,19 @@ def ConcatOp : VariadicRTLOp<"concat"> {
   ];
 }
 
+def Extractable
+  : Type<CPred<"$_self.isa<IntegerType>() ||"
+               "$_self.isa<ArrayType>()">,
+         "a type which can be sliced">;
+
 // Extract a range of bits from the specified input.
 def ExtractOp : RTLOp<"extract", [NoSideEffect]> {
```

Comment:
I don't understand the goal of this - it looks like you want to be able to slice values out of value semantic arrays?

Why not keep arrays as lvalues (aka, inouts) that are processed with array index op?  This more closely follows verilog semantics and doesn't have a potential blowup problem if the array has to be scalarized.

Separate from the "why do we want this" question, I don't think this should be overlaid on top of extractop.  We should have two different operations for extracting slices of bits out of an integer from array slices.

Reply:
I realize now that this was probably worded too strongly.  I'm not sure what the tradeoffs are here, but I also realize we may not have a choice.  SVerilog supports structs as the type ports, and if the elements of those structs can be arrays, then we need value semantic arrays.

Reply:
I must not understand `inout`. I thought that it was intended to support bidirectional wires? If so, I don't see any relationship between value semantic arrays and `inout`. From some of your code and comments, I think you're treating `inout` like a pointer?

SVerilog (and I think it makes sense for the RTL dialect too) supports arbitrary (size, shape, and dimension) arrays as module parameters. For modeling, value type on those makes sense to me since they can be mutated combinationally (and whatever gets fed by the mutated version sees the mutated version). Or they can get stored. Or some combination thereof.

In terms of using the extract op, I was debating that myself. The reason I went with extract is since in hardware, there's no difference between arrays of bits and integers. Argument against is that we probably don't want to intro that low level hardware lack of typing to the RTL dialect. It's reasonable to introduce a `slice` operation to do the same.